[FEATURE] [Certif] Afficher le bouton et la page pour envoyer des invitations (PIX-134)

lego-technix opened the pull request:
## :unicorn: Problème

Avec l'arrivée des rôles dans Pix Certif, il faut une nouvelle page pour pouvoir saisir les adresses email pour envoyer les invitations.

## :robot: Proposition

* Ajout d'un bouton “Inviter un membre” (ce bouton est à droite du titre de la page, et donc accessible peu importe l’onglet que l’on visualise “Membres” ou “Invitations”)
  * ce bouton n’est visible que si le user a le rôle ADMIN
  * ce bouton est aligné à droite de la même manière que le bouton « Changer de référent » quand il s'affiche
* Au clic, on arrive sur une nouvelle page, iso Pix Orga
  * du texte (idem Pix Orga)
  * un champ de saisie, où il est possible de saisir plusieurs adresses email en une fois, séparées par une virgule
  * 2 boutons d’action : Annuler et Inviter

## :rainbow: Remarques

### Nommage

Dans cette PR, les fichiers liés au formulaire d'invitation ont pour base le nom `invite`, alors que dans Pix Orga les fichiers correspondants ont pour base les noms `create` et `new`. Mais `create` et `new` font penser à la création de membres, alors qu'en réalité on ne fait que créer des invitations qui potentiellement seulement donneront lieu à de nouveaux membres.

### Codage

Le premier commit de cette PR sert à mettre en place à 100% la logique de [_nested route_](https://guides.emberjs.com/release/routing/defining-your-routes/#toc_nested-routes) pour la route `/equipe`.

### Nouvelle seed

Un nouveau centre de certification `Accèssovolt` est créé avec cette PR. Ce centre de certification permet de tester fonctionnellement le cas où un utilisateur  `james-paledroits@example.net` est _Admin_ dans un centre et _Membre_ dans un autre.

## :100: Pour tester

1. Se connecter dans [Pix Certif](https://certif-pr7173.review.pix.fr/) avec un compte ayant le rôle d'_Admin_ dans un centre de certification et ayant un rôle _Membre_ dans un autre centre de certification (par exemple `james-paledroits@example.net` est _Admin_ dans le centre de certification `Accèsorium` et _Membre_ dans le centre de certification `Accèssovolt`)
2. Constater que sur la page d'équipe du centre de certification où l'utilisateur est _Admin_ un nouveau bouton `Inviter un membre` est affiché à droite
3. Cliquer sur ce bouton et constater qu'on est redirigé vers une nouvelle page `Inviter un membre`
4. Se rendre sur le centre de certification où l'utilisateur n'est pas _Admin_
5. Constater que sur la page d'équipe du centre de certification le nouveau bouton `Inviter un membre` n'est pas affiché

Faire cette manipulation à la fois sur une organisation « CLEA » et sur une organisation non-« CLEA ». La colonne  	`Référent` doit s'afficher pour les organisations « CLEA ».

lego-technix committed:
refactor(certif): setup nested route for team/list

https://guides.emberjs.com/release/routing/defining-your-routes/

Co-authored-by: BKsensei <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>
Co-authored-by: Yoan de LUCA <yoan.de-luca@pix.fr>

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr7173.review.pix.fr)
  * [App (.org)](https://app-pr7173.review.pix.org)
  * [Orga (.fr)](https://orga-pr7173.review.pix.fr)
  * [Orga (.org)](https://orga-pr7173.review.pix.org)
  * [Certif (.fr)](https://certif-pr7173.review.pix.fr)
  * [Certif (.org)](https://certif-pr7173.review.pix.org)
  * [1D](https://1d-pr7173.review.pix.fr)
  * [Admin](https://admin-pr7173.review.pix.fr)
  * [API](https://api-pr7173.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr7173.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr7173/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr7173/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr7173/environment)

lego-technix commented:
À FAIRE : 
* Reprendre le layout des 2 boutons pour que les 2 boutons soient alignés à droite
* Corriger une régression qui a fait disparaître la 3ème colonne « Référent »
* Voir si on peut transformer le test d'acceptance 'it displays an invite a member button' en test d'intégration

lego-technix commented on certif/app/router.js:
```diff
@@ -40,6 +40,7 @@ Router.map(function () {
         this.route('members', { path: '/membres' });
         this.route('invitations');
       });
+      this.route('invite', { path: '/inviter' });
```
TODO : Actuellement cette route n'a pas son accès protégé pour que seuls les admins y accèdent. C'est à corriger.

lego-technix committed:
refactor(certif): add missing route auth test

and also redirect to the root "authenticated.team" route which is more lasting
because it is a root path which will always be here while other sub/nested
routes might more easily evolve.

lego-technix committed:
refactor(certif): typo: prefer correct spelling

reibecca commented on certif/app/templates/authenticated/team/invite.hbs:
```diff
@@ -0,0 +1,25 @@
+    {{t "pages.team-invite.title"}}
+  </h1>
+
+  <p>
+    {{t "pages.team-invite.email-requirement"}}<br />
+    {{t "pages.team-invite.several-email-requirement"}}
+  </p>
+
+  <p>
+    {{t "pages.team-invite.invited-members"}}
+  </p>
```
**chipotage :** changer la font du texte présent au dessus du formulaire d'invitation en 14px au lieu de 16px (possibilité d'utiliser le placeholder  `@extend %pix-body-s`

reibecca commented on certif/translations/fr.json:
```diff
@@ -887,6 +888,14 @@
+      "title": "Inviter un membre",
+      "email-requirement": "Saisissez ici l'adresse e-mail du membre que vous souhaitez inviter.",
+      "input-label": "Adresse(s) e-mail",
+      "invite-button": "Inviter un membre",
```
**chipotage :** modifier le wording du bouton pour envoyer une invitation en écrivant simplement "inviter" comme sur pix orga, comme on aura la possibilité d'inviter un ou plusieurs membres

reibecca reviewed (approved):
✅ Testé fonctionnellement sur Chrome et Safari

Thomasevano commented on certif/tests/acceptance/team_test.js:
```diff
@@ -102,6 +105,34 @@ module('Acceptance | authenticated | team', function (hooks) {
       });
     });
 
+    module('when visiting "/equipe/invitations" unauthorized URL', function () {
```
```suggestion
    module('when visiting "/equipe/invitations" as an unauthorized user', function () {
```

Je sais pas si c'est la bonne formulation, mais pour moi c'est l'utilisateur qui a pas les droits d'accéder a l'URL

Thomasevano commented on certif/tests/acceptance/team_test.js:
```diff
@@ -102,6 +105,34 @@ module('Acceptance | authenticated | team', function (hooks) {
+      });
+    });
+
+    module('when visiting "/equipe/inviter" unauthorized URL', function () {
```
```suggestion
    module('when visiting "/equipe/inviter" as an unauthorized user', function () {
```

Steph0 commented:
# Tests funcs ❌ 

Se connecter dans [Pix Certif](https://certif-pr7173.review.pix.fr/) avec un compte ayant le rôle d'Admin dans une organisation (par exemple james-paledroits@example.net) ✅ 

Constater que dans cette équipe un nouveau bouton Inviter un membre est affiché à droite✅ 

Cliquer sur ce bouton et constater qu'on est redirigé vers une nouvelle page Inviter un membre✅ 
![image](https://github.com/1024pix/pix/assets/170271/49f2e0ae-39f0-4c38-8b6f-d0f742821671)

Se déconnecter ✅

Se connecter dans [Pix Certif](https://certif-pr7173.review.pix.fr/) avec un compte ayant le rôle de Membre dans une organisation (par exemple marc-alex-terrieur@example.net)✅
Constater que dans cette équipe le nouveau bouton Inviter un membre n'est affiché✅

Faire cette manipulation à la fois sur une organisation « CLEA » et sur une organisation non-« CLEA ». La colonne Référent doit s'afficher pour les organisations « CLEA ».

CLEA ✅ 
![image](https://github.com/1024pix/pix/assets/170271/fd1eb5c5-df01-4ad5-98f3-7ca867b794f6)
![image](https://github.com/1024pix/pix/assets/170271/fccf8e79-88fa-4d0f-a492-c9d1e37f4235)

PAS CLEA ❌ 

![image](https://github.com/1024pix/pix/assets/170271/947a198b-0476-4203-a57c-f2fc0d0705ee)
![image](https://github.com/1024pix/pix/assets/170271/2127dd4c-3bb0-48c1-b967-c9b8005c026f)
![image](https://github.com/1024pix/pix/assets/170271/dbec2316-5eea-4762-a0c8-b9d7a9b8a486)


J'ai peut-être loupé quelque chose dans ma compréhension mais le PAS CLEA devrait donc pas avoir le bouton ?

-----------------------------
Plusieurs orga, differents droits ❌

Aussi j'ai retiré les droits Admin sur l'orga 7001 (la SCO managing student) et d'un coup le bouton a disparu partout, par exemple sur Accessorium, alors que sur orga il est bien Admin

![image](https://github.com/1024pix/pix/assets/170271/7145bd9f-90a2-41c2-a38d-b95efc7c449c)
![image](https://github.com/1024pix/pix/assets/170271/bc332c0a-ad40-4d7b-b888-63ce4ad569eb)
![image](https://github.com/1024pix/pix/assets/170271/2aef04e1-c710-4c62-ab2b-36aba8389598)

Steph0 commented on certif/app/components/team/invite-form.hbs:
```diff
@@ -0,0 +1,26 @@
+{{! template-lint-disable require-input-label }}
```
Pourquoi ne pas mettre les label (cela remonte aussi dans WAVE niveau accessibilité) ?

lego-technix replied:
Ce `{{! template-lint-disable require-input-label }}` a été retiré, il s'agissait d'une propagation involontaire par copier-coller.

Steph0 commented on certif/app/templates/authenticated/team/invite.hbs:
```diff
@@ -0,0 +1,25 @@
+{{page-title (t "pages.team-invite.title")}}
+
+<div class="team-invite-page">
```
Alors sous-couvert de ma bonne compréhension de la petite formation faite avec @frinyvonnick ; je crois qu'en BEM il faut pas rappeler le nom du fichier parent (ici `/team` par rapport à `/invite`).
Pinaillage : je trouve aussi que `page` est pas forcement utile

```suggestion
<div class="invite">
```

PLus d'infos ici : https://1024pix.atlassian.net/wiki/spaces/DC/pages/3845292101/BEM+Review

reibecca replied:
je ne suis pas très d'accord, avoir la classe principale avec le mot 'page' dedans est très répandu dans le code de Pix sur les templates, ça permet une meilleure compréhension je trouve

reibecca replied:
je ne suis pas très d'accord, avoir la classe principale avec le mot 'page' dedans est très répandu dans le code de Pix sur les templates, ça permet une meilleure compréhension je trouve

Steph0 replied:
@reibecca je n'avais pas forcèment regardé, mais effectivement sur le `-page` je trouve 218 occurences dans les hbs 😄

Steph0 commented on certif/app/templates/authenticated/team/invite.hbs:
```diff
@@ -0,0 +1,25 @@
+{{page-title (t "pages.team-invite.title")}}
+
+<div class="team-invite-page">
+  <h1 class="page__title page-title">
```
Retour BEM, le nom du bloc (`team-invite-page`) doit être nommé entièrement. N'y aurait-il pas aussi un truc redondant d'avoir à la fois (`page__title` et `page-title`) ?

```suggestion
  <h1 class="team-invite-page__title">
```

aceol replied:
@Steph0  Les deux sont contradictoires non? Pourquoi la 1ere?
> je crois qu'en BEM il faut pas rappeler le nom du parent.

> Retour BEM, le parent doit être nommé entièrement

Steph0 replied:
Vu ensemble, j'ai modifié les textes de mes commentaires pour que ça soit + compréhensible ^^'

lego-technix replied:
@Steph0 bien vu cette classe `page__title` problématique ! En fait il n'y a pas de style associé dans Pix Certif. C'est une classe qui vient de Pix Orga, et qui existe dans Pix Orga, mais qui n'existe pas dans Pix Certif. Cette classe s'est propagée dans Pix Certif par copier-coller. Je l'ai enlevée en poussant un commit fixup.

Steph0 commented on certif/app/templates/authenticated/team/list.hbs:
```diff
@@ -0,0 +1,67 @@
+      </div>
+    </div>
+  </div>
+{{/if}}
+
+{{#if this.shouldDisplayNavbarSection}}
```
Il n'y a en théorie pas de cas possible où on aurait les deux simultanement. Applicable aussi à `{{#if this.shouldDisplayInviteMemberButton}}`

```suggestion
{{else if this.shouldDisplayNavbarSection}}
```

aceol replied:
Pourquoi pas les 2 simultanéments? Il ne semble pas s'opposer

Steph0 replied:
Punaise j'ai pas spotté le bon endroit du fichier..... 

Ce commentaire est pour `{{#if this.shouldDisplayNoRefererSection}}` et `{{#if this.shouldDisplayUpdateRefererButton}}`  pas sur la navbarsection, c'est a dire suggestion pour mettre `({{#else if shouldDisplayUpdateRefererButton}}`

Pas de NoRefererSection et DisplayUpdateRefererButton simultanément (par contre on peut très bien avoir la navbar xD)

lego-technix replied:
Je ne suis pas certain de bien comprendre car `{{#if this.shouldDisplayUpdateRefererButton}}` et `{{#if this.shouldDisplayNoRefererSection}}` ne sont pas au même niveau dans l'arborescence DOM de ce fragment HTML, et donc je ne pense pas qu'on puisse les liés avec un `else`.

Steph0 replied:
J'ai relu tout cela attentivement, on dirait bien que c'est moi qui me suis mélangé les pinceaux 😬

Steph0 commented on certif/app/components/team/invite-form.hbs:
```diff
@@ -0,0 +1,26 @@
+{{! template-lint-disable require-input-label }}
+<form {{on "submit" @onSubmit}} class="form" ...attributes>
```
Il n'y a pas de fichier de test sur ce composant

lego-technix replied:
@Steph0 bien vu ! Le fichier de test `certif/tests/integration/components/team/invite-form_test.js` correspondant a été ajouté.

Steph0 reviewed (changes requested):
Sur les tests funcs si je me suis plantés sur les KO cela m'intéresse d'avoir une explication personnalisée ! :)

Je mets Request changes à cause de ces tests car si ce sont de vrais KO je me dis que la PR ne peut pas passer en l'état

reibecca commented:
> J'ai peut-être loupé quelque chose dans ma compréhension mais le PAS CLEA devrait donc pas avoir le bouton ?

@Steph0 si PAS CLEA, alors le bouton concernant le réferent ne doit pas s'afficher. 
On veut le bouton pour inviter un nouveau membre quelque soit le type de l'organisation, donc c'est normal qu'il soit toujours présent, c'est ce qu'on veut 🙂

reibecca commented:
> Plusieurs orga, differents droits ❌

> Aussi j'ai retiré les droits Admin sur l'orga 7001 (la SCO managing student) et d'un coup le bouton a disparu partout, par exemple sur Accessorium, alors que sur orga il est bien Admin

@Steph0  La règle sur Pix Certif est que seul un admin peut inviter d'autres membres dans un centre de certification. Du coup si plus d'admin, plus de possibilité d'effectuer cette action.

Par contre, il me semble que sur Pix Orga, il y a une règle qui fait qu'on ne peut pas avoir d'Orga sans admin (à vérifier je suis pas sur à 100%), ça expliquerait pourquoi l'admin a toujours son rôle coté Orga

igorissen commented:
> > Plusieurs orga, differents droits ❌
> 
> > Aussi j'ai retiré les droits Admin sur l'orga 7001 (la SCO managing student) et d'un coup le bouton a disparu partout, par exemple sur Accessorium, alors que sur orga il est bien Admin
> 
> @Steph0 La règle sur Pix Certif est que seul un admin peut inviter d'autres membres dans un centre de certification. Du coup si plus d'admin, plus de possibilité d'effectuer cette action.
> 
> Par contre, il me semble que sur Pix Orga, il y a une règle qui fait qu'on ne peut pas avoir d'Orga sans admin (à vérifier je suis pas sur à 100%), ça expliquerait pourquoi l'admin a toujours son rôle coté Orga

@Steph0 le rôle sur Pix Certif n'est pas lié au rôle sur Pix Orga, c'est 2 tables distinctes. Donc si tu as retirés que celui de Pix Certif alors c'est normal que sur Pix Orga l'utilisateur a toujours son rôle admin (j'espère avoir bien compris la conversion 😅).

- table `memberships` pour la partie Pix Orga. Et comme @reibecca l'a signalé sur Pix Orga il doit y avoir toujours un admin. Cette condition n'est pas encore disponible sur Pix Certif (enfin je pense j'étais en congé 😅).
- table `certification-center-memberships` pour la partie Pix Certif

igorissen commented on certif/app/routes/authenticated/team.js:
Ce code aurait du rester ici.
- Il permet de récupérer la liste des membres et de le récupérer via this.modelFor(...) depuis la route enfant
- Idem pour la liste des invitations qui n'est pas censé être au niveau de la route `/equipe/members`

igorissen replied:
Je viens de voir qu'il a été déplacé dans `team.list` parent de `team.list.members` et `team.list.invitations`

aceol commented:
⚠️ il y a tout pleins de commit de fixup

igorissen commented on certif/tests/unit/controllers/authenticated/team/invite_test.js:
```diff
@@ -0,0 +1,22 @@
+  let controller;
+
+  hooks.beforeEach(function () {
+    controller = this.owner.lookup('controller:authenticated/sessions/new');
```
```suggestion
    controller = this.owner.lookup('controller:authenticated/team/invite');
```

igorissen commented on certif/tests/unit/controllers/authenticated/team/invite_test.js:
```diff
@@ -0,0 +1,22 @@
+    await controller.send('cancel');
+
+    // then
+    assert.ok(controller.router.transitionTo.calledWith('authenticated.sessions.list'));
```
```suggestion
    assert.ok(controller.router.transitionTo.calledWith('authenticated.team.invite'));
```

lego-technix replied:
Cette ligne avait bien besoin d'être corrigée, merci. Mais la redirection dans le controller est faite vers `authenticated.team` et voila donc pourquoi je viens de pousser un fixup commit qui modifie l'assert pour qu'il soit comme suit : 

```javascript
assert.ok(controller.router.transitionTo.calledWith('authenticated.team'));
```

igorissen commented on certif/tests/acceptance/team_test.js:
```diff
@@ -102,6 +105,34 @@ module('Acceptance | authenticated | team', function (hooks) {
     });
 
+    module('when visiting "/equipe/invitations" as an unauthorized user', function () {
+      test('redirects to "/equipe" URL', async function (assert) {
```
```suggestion
      test('redirects to "/equipe/membres" URL', async function (assert) {
```

igorissen commented on certif/tests/acceptance/team_test.js:
```diff
@@ -102,6 +105,34 @@ module('Acceptance | authenticated | team', function (hooks) {
+    });
+
+    module('when visiting "/equipe/inviter" as an unauthorized user', function () {
+      test('redirects to "/equipe" URL', async function (assert) {
```
```suggestion
      test('redirects to "/equipe/membres" URL', async function (assert) {
```

igorissen reviewed (commented):
Comme signaler, il manque un test d'intégration sur le composant `components/team/invite-form`.

igorissen commented:
> ⚠️ il y a tout pleins de commit de fixup

Oui, c'est normal. On a décider avec l'équipe de laisser les fixup pour ceux qui font les reviews pour voir ce qui a été modifié après le commentaire. Une fois que tout est résolu, on fait le fixup sur les commits associés puis on lance le processus de merge.

aceol commented:
> > ⚠️ il y a tout pleins de commit de fixup
> 
> Oui, c'est normal. On a décider avec l'équipe de laisser les fixup pour ceux qui font les reviews pour voir ce qui a été modifié après le commentaire. Une fois que tout est résolu, on fait le fixup sur les commits associés puis on lance le processus de merge.
OK ca marche!
(du coup, c'est pas evident quand on fait une nouvelle review commit par commit)

er-lim committed:
fix(certif): update current membership when changing certification

center

Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>

lego-technix committed:
feat(api): add new certification center in seeds

lego-technix commented:
> Sur les tests funcs si je me suis plantés sur les KO cela m'intéresse d'avoir une explication personnalisée ! :)
> 
> Je mets Request changes à cause de ces tests car si ce sont de vrais KO je me dis que la PR ne peut pas passer en l'état

@Steph0 bien vu pour les tests fonctionnels ! Le problème que tu as remonté a été corrigé avec un commit fixup (merci @er-lim).

Et il y a maintenant dans les seeds `team-acces` un nouveau centre de certification `Accèssovolt` pour l'utilisateur `james-paledroits@example.net`. Cet utilisateur est maintenant _Admin_ dans  le centre de certification `Accèsorium` et _Membre_ dans le centre de certification `Accèssovolt`. Cela devrait éviter à l'avenir que l'on passe à côté de ce genre de problème.

Enfin, j'ai modifié la procédure _Pour tester_ de manière à y inclure le point que tu as testé et qui était absent.

yoandl commented:
Tests sur Firefox  ✅  : 
tous les tests semblent ok, le bouton "inviter les membres" est bien présent pour un utilisateur admin (clea /non-clea et la colonne Référent apparait bien pour les centres CLEA .

yoandl reviewed (approved):
Tests sur Firefox ✅ :
tous les tests semblent ok, le bouton "inviter les membres" est bien présent pour un utilisateur admin (clea /non-clea) et la colonne Référent apparait bien pour les centres CLEA .

lego-technix commented:
Toutes vos remarques ont été traitées. Vous êtes invités à marquer comme « Resolved » toutes les conversations qui le sont et à prévenir si un point avait été corrigé de manière non satisfaisante ou oublié. Merci !

Steph0 reviewed (approved):
Après revue, tout me semble OK désormais !

lego-technix committed:
feat(certif): add invite a member button and page

Co-authored-by: BKsensei <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>
Co-authored-by: Emmanuelle Bonnemay <emmanuelle.bonnemay@pix.fr>
Co-authored-by: Yoan de LUCA <yoan.de-luca@pix.fr>
